tests/test_alembic.py

```diff
@@ -0,0 +1,169 @@
+    '''
+    models_homedir = str(tmpdir.mkdir('models'))
+    subprocess.check_call(['sqlite3', os.path.join(models_homedir, 'svs.sqlite'), '.databases'])
+    engine = make_engine(models_homedir)
+    Base.metadata.create_all(bind=engine, checkfirst=False)
+    session = sessionmaker(engine)()
+    models_schema = get_schema(session)
+
+    alembic_homedir = str(tmpdir.mkdir('alembic'))
+    subprocess.check_call(['sqlite3', os.path.join(alembic_homedir, 'svs.sqlite'), '.databases'])
+    session = sessionmaker(make_engine(alembic_homedir))()
+    alembic_config = conftest._alembic_config(alembic_homedir)
+    upgrade(alembic_config, 'head')
+    alembic_schema = get_schema(session)
+
+    # The initial migration creates the table 'alembic_version', but this is
+    # not present in the schema created by `Base.metadata.create_all()`.
+    alembic_schema = {k: v for k, v in alembic_schema.items()
+                      if k[2] != 'alembic_version'}
+
+    assert_schemas_equal(alembic_schema, models_schema)
+
+
+@pytest.mark.parametrize('migration', ALL_MIGRATIONS)
+def test_alembic_migration_upgrade(alembic_config, config, migration):
+    # run migrations in sequence from base -> head
+    for mig in list_migrations(alembic_config, migration):
+        upgrade(alembic_config, mig)
+
+
+@pytest.mark.parametrize('migration', ALL_MIGRATIONS)
+def test_alembic_migration_downgrade(alembic_config, config, migration):
+    # upgrade to the parameterized test case ("head")
+    upgrade(alembic_config, migration)
+
+    # run migrations in sequence from "head" -> base
+    migrations = list_migrations(alembic_config, migration)
+    migrations.reverse()
+
+    for mig in migrations:
+        downgrade(alembic_config, mig)
+
+
+@pytest.mark.parametrize('migration', ALL_MIGRATIONS)
+def test_schema_unchanged_after_up_then_downgrade(alembic_config,
+                                                  tmpdir,
+                                                  migration):
+    migrations = list_migrations(alembic_config, migration)
+
+    if len(migrations) > 1:
+        target = migrations[-2]
+        upgrade(alembic_config, target)
+    else:
+        # The first migration is the degenerate case where we don't need to
+        # get the database to some base state.
+        pass
+
+    session = sessionmaker(make_engine(str(tmpdir.mkdir('original'))))()
+    original_schema = get_schema(session)
+
+    upgrade(alembic_config, '+1')
+    downgrade(alembic_config, '-1')
+
+    session = sessionmaker(make_engine(str(tmpdir.mkdir('reverted'))))()
+    reverted_schema = get_schema(session)
+
+    # The initial migration is a degenerate case because it creates the table
+    # 'alembic_version', but rolling back the migration doesn't clear it.
+    if len(migrations) == 1:
+        reverted_schema = {k: v for k, v in reverted_schema.items() if k[2] != 'alembic_version'}
+
+    assert_schemas_equal(reverted_schema, original_schema)
```

Comment:
we reviewed this over in https://github.com/freedomofpress/securedrop/pull/3260, so this logic is good imho